### PR TITLE
[Core] Adding methods for custom IntegrationPointsArray

### PR DESCRIPTION
**Pull request by loumalouomega**
This will allow to more easily consider non static defined integration methods

Part of #6734

### CODE REVIEW

**Comment by pablobecker** [comment]
IMO this should be adressed by the design or technical comitee rather than the implementation one. there's not much to discuss regarding implemantion

**Comment by pablobecker** [comment]
(not pinging the design comittee until I discuss this with the rest of the @KratosMultiphysics/implementation-committee )

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -1209,6 +1209,20 @@ class Geometry
         return result;
     }
 
+    /**
+     * @brief It returns a vector that is normal to its corresponding geometry in a given integration method
```
Comment (GuillermoCasas) [inline kratos/geometries/geometry.h:1553]: should 'in a given integration method' not be 'in a given integration point'?

File: kratos/geometries/geometry.h [hunk 2]
```diff
@@ -1209,6 +1209,20 @@ class Geometry
...
+     * @param IntegrationPointIndex index of integration point which jacobians has to be calculated in it.
+     * @param rIntegrationPointsArray The integration method which jacobians has to be calculated in its integration points.
+     * @return The normal in the given integration method
```
Comment (GuillermoCasas) [inline kratos/geometries/geometry.h:1556]: I suggest 'the normal on the given integration point'

File: kratos/geometries/geometry.h [hunk 3]
```diff
@@ -1209,6 +1209,20 @@ class Geometry
         return result;
     }
 
+    /**
+     * @brief It returns a vector that is normal to its corresponding geometry in a given integration method
+     * @param IntegrationPointIndex index of integration point which jacobians has to be calculated in it.
+     * @param rIntegrationPointsArray The integration method which jacobians has to be calculated in its integration points.
```
Comment (GuillermoCasas) [inline kratos/geometries/geometry.h:1555]: This sentence is wrong. Woud just 'Array containing the integration points corresponding to a particular integration method' work?

Comment (loumalouomega) [reply kratos/geometries/geometry.h:1555]: I will correct those typos

File: kratos/geometries/geometry.h [hunk 4]
```diff
@@ -2027,6 +2055,22 @@ class Geometry
...
     }
 
+    /**
+     * @brief This method provides the global coordinates corresponding to the given integration method.
+     * @param rResult The array containing the global coordinates corresponding to the given integration method.
+     * @param IntegrationPointIndex index of integration point which jacobians has to be calculated in it.
+     * @param rIntegrationPointsArray The integration method which jacobians has to be calculated in its integration points.
+     * @return An array containing the global coordinates corresponding to the given integration method.
+     */
+    virtual CoordinatesArrayType& GlobalCoordinates(
```
Comment (GuillermoCasas) [inline kratos/geometries/geometry.h:2065]: why virtual? Is this not general enough when the main function GlobalCoordinates is already virtual?

File: kratos/geometries/geometry.h [hunk 5]
```diff
@@ -2445,6 +2489,30 @@ class Geometry
...
 
+    /**
+     * @brief Jacobians for given  method.
+     * @details This method calculate jacobians matrices in all integrations points of given integration method.
+     * @param rIntegrationPointsArray The integration method which jacobians has to be calculated in its integration points.
+     * @return JacobiansType a Vector of jacobian matrices \f$ J_i \f$ where \f$ i=1,2,...,n \f$ is the integration point index of given integration method.
+     * @see DeterminantOfJacobian
+     * @see InverseOfJacobian
+     */
+    virtual JacobiansType& Jacobian(
```
Comment (pablobecker) [inline kratos/geometries/geometry.h:2500]: singular?

**Review by GuillermoCasas — DISMISSED** [review]
On behalf of the @KratosMultiphysics/implementation-committee , we have looked at your implementation and it adds valuable general functionalities for the geometry. Since the implementation is just based on calling existing functions, and since not every function in the core (by no means) is currently being tested, we agreed to approve this PR. Nonetheless, it would be nice to have a test with an element(s) that make use of these functions in a nontrivial way. We encourage you to add such a test if you can.

**Comment by GuillermoCasas** [comment]
> On behalf of the @KratosMultiphysics/implementation-committee , we have looked at your implementation and it adds valuable general functionalities for the geometry. Since the implementation is just based on calling existing functions, and since not every function in the core (by no means) is currently being tested, we agreed to approve this PR. Nonetheless, it would be nice to have a test with an element(s) that make use of these functions in a nontrivial way. We encourage you to add such a test if you can.

We just realised that you are adding such a test in #6734. Good job!

**Comment by philbucher** [comment]
I think @pooyan-dadvand has some comments, please wait with merging

**Comment by loumalouomega** [comment]
@pooyan-dadvand may I proceed?

**Comment by loumalouomega** [comment]
@pooyan-dadvand should I proceed?

**Comment by RiccardoRossi** [comment]
apparently a leak has been found which is related to #6750  so please wait until that is on.

@pooyan-dadvand  can u also comment on that?

**Comment by loumalouomega** [comment]
> 
> 
> apparently a leak has been found which is related to #6750 so please wait until that is on.
> 
> @pooyan-dadvand can u also comment on that?

what kind of leak?

**Comment by loumalouomega** [comment]
The "leak" is fixed in https://github.com/KratosMultiphysics/Kratos/pull/6926

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Maybe I didn't understand well the problem but I feel we are not in the correct direction with this change. Let revise it again:

1. We wanted to have different integration rules in the geometries
2. Geometry will store these integration points (as a static or normal member)
3. We already have Normal, Global Coordinates, etc over gauss points
4. Now we add such methods for a given array of integration points

So my question is, why we are not simply passing the new integration method enum and using the array stored in the geometry rather than duplicating the interface for passing an array which we have it inside the goemetry?

**Comment by loumalouomega** [comment]
> 
> 
> Maybe I didn't understand well the problem but I feel we are not in the correct direction with this change. Let revise it again:
> 
>     1. We wanted to have different integration rules in the geometries
> 
>     2. Geometry will store these integration points (as a static or normal member)
> 
>     3. We already have Normal, Global Coordinates, etc over gauss points
> 
>     4. Now we add such methods for a given array of integration points
> 
> 
> So my question is, why we are not simply passing the new integration method enum and using the array stored in the geometry rather than duplicating the interface for passing an array which we have it inside the goemetry?

The problem is that we can just define 10 integration methods. This PR has as objective to give more flexibility in order to consider integration methods not previously added and defined in the geometries (this has a limit opf 10 as I said and requires to modify very important and not supposed to be modified so often classes, the geometries). This PR is required in order to add the Newton-Cotes integration methods from #6734

**Comment by pooyan-dadvand** [comment]
Still, I am not following you. Why we can only define 10 integration methods?

To my understanding, if we add any new integration method to the enum in `geometry_data.h` it should extend that number:

```cpp
    enum IntegrationMethod {
        GI_GAUSS_1,
        GI_GAUSS_2,
        GI_GAUSS_3,
        GI_GAUSS_4,
        GI_GAUSS_5,
        GI_EXTENDED_GAUSS_1,
        GI_EXTENDED_GAUSS_2,
        GI_EXTENDED_GAUSS_3,
        GI_EXTENDED_GAUSS_4,
        GI_EXTENDED_GAUSS_5,
        GI_A_NEW_AFTER_10_INTEGRATION,
        NumberOfIntegrationMethods
    };
```
of course, you cannot ask this new integration rule from the geometries which does not implement it. (We may ward it with a debug error)

Meanwhile, @tteschemacher  is working on a new IntegrationInfo class which separates the integration method itself and the order of integration that may be very useful in this case.

**Comment by loumalouomega** [comment]
> Still, I am not following you. Why we can only define 10 integration methods?
> 
> To my understanding, if we add any new integration method to the enum in `geometry_data.h` it should extend that number:
> 
> ```c++
>     enum IntegrationMethod {
>         GI_GAUSS_1,
>         GI_GAUSS_2,
>         GI_GAUSS_3,
>         GI_GAUSS_4,
>         GI_GAUSS_5,
>         GI_EXTENDED_GAUSS_1,
>         GI_EXTENDED_GAUSS_2,
>         GI_EXTENDED_GAUSS_3,
>         GI_EXTENDED_GAUSS_4,
>         GI_EXTENDED_GAUSS_5,
>         GI_A_NEW_AFTER_10_INTEGRATION,
>         NumberOfIntegrationMethods
>     };
> ```
> 
> of course, you cannot ask this new integration rule from the geometries which does not implement it. (We may ward it with a debug error)
> 

That's the problem, the requirement to extend manually the IntegrationMethod enum. The methods here presented avoid that requirement so it can be defined on the fly without previous modification.

> Meanwhile, @tteschemacher is working on a new IntegrationInfo class which separates the integration method itself and the order of integration that may be very useful in this case.

That's nice

**Comment by tteschemacher** [comment]
>     Meanwhile, @tteschemacher is working on a new IntegrationInfo class which separates the integration method itself and the order of integration that may be very useful in this case.
> 
> That's nice

It is in #7016 .

Do you need those interfaces? Isn't it easier to call those functions from outside?

**Comment by loumalouomega** [comment]
> > ```
> > Meanwhile, @tteschemacher is working on a new IntegrationInfo class which separates the integration method itself and the order of integration that may be very useful in this case.
> > ```
> > 
> > 
> > That's nice
> 
> It is in #7016 .
> 
> Do you need those interfaces? Isn't it easier to call those functions from outside?

I don't understand

**Comment by tteschemacher** [comment]
> I don't understand

You are never accessing any private function or memory. Thus, you can just do the loops through the integration points externally. May be easier.

**Comment by loumalouomega** [comment]
> > I don't understand
> 
> You are never accessing any private function or memory. Thus, you can just do the loops through the integration points externally. May be easier.

But then you need to define manually each time...

**Comment by RiccardoRossi** [comment]
it looks like the conversation is stuck. Shall we close this one or is it still considered active?

if it is still active please make a resume of the open points and on how the reviewer's concerns were addressed

**Comment by loumalouomega** [comment]
I think this can be retaken

**Comment by rubenzorrilla** [comment]
> I think this can be retaken @KratosMultiphysics/technical-committee

As it was demanded by  @RiccardoRossi in his previous comment, in a nutshell, the question is "whether we allow or not passing custom integration rules (that do not belong to the `IntegrationMethod` `enum`)". Am I wrong @loumalouomega?

**Comment by loumalouomega** [comment]
> > I think this can be retaken @KratosMultiphysics/technical-committee
> 
> As it was demanded by @RiccardoRossi in his previous comment, in a nutshell, the question is "whether we allow or not passing custom integration rules (that do not belong to the `IntegrationMethod` `enum`)". Am I wrong @loumalouomega?

Yes, it is was something required by @Vahid-Galavi to use custom integration schemes

**Review by pooyan-dadvand — COMMENTED** [review]
@KratosMultiphysics/technical-committee reviewed this changes again and arrive to the conclusion that these interfaces are not really required for these reasons:

1. One can make the same loop outside and calling the original functions passing the coordinate of the integration points (as suggested by @tteschemacher)
2. In case that we want to use a different integration rule than the ones already in the given geometry, we can pass them as local coordinates one by one in the integration loop.

**Pull request closed without merging**